File: app/src/main/java/de/sysitprep/app/ui/quiz/ResultViewModel.kt
package de.sysitprep.app.ui.quiz

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import de.sysitprep.app.ExamForgeApp
import de.sysitprep.app.data.model.QuizResult
import de.sysitprep.app.data.model.QuestionResult
import de.sysitprep.app.data.model.Session
import de.sysitprep.app.data.repository.QuestionRepository
import kotlinx.coroutines.launch

class ResultViewModel : ViewModel() {

    private val repository: QuestionRepository by lazy { ExamForgeApp.instance.repository }

    private val _quizResult = MutableLiveData<QuizResult?>()
    val quizResult: LiveData<QuizResult?> = _quizResult

    fun loadResult(sessionId: Long) {
        viewModelScope.launch {
            val session = repository.getSessionById(sessionId) ?: return@launch
            val answers = repository.getAnswersForSession(sessionId)

            val questionResults = answers.map { answer ->
                val question = repository.getQuestionById(answer.questionId)
                QuestionResult(
                    questionId = answer.questionId,
                    questionText = question?.questionText ?: "",
                    options = question?.options ?: emptyList(),
                    correctIndices = question?.correctAnswerIndices ?: emptyList(),
                    selectedIndices = answer.selectedIndices,
                    isCorrect = answer.isCorrect,
                    explanation = question?.explanation ?: ""
                )
            }

            val total = questionResults.size
            val correct = questionResults.count { it.isCorrect }
            val scorePercent = if (total > 0) (correct * 100 / total) else 0
            val passed = scorePercent >= 50
            val timeSeconds = session.durationSeconds ?: 0

            _quizResult.value = QuizResult(
                sessionId = sessionId,
                totalQuestions = total,
                correctAnswers = correct,
                wrongAnswers = total - correct,
                skippedAnswers = questionResults.count { it.selectedIndices.isEmpty() },
                scorePercent = scorePercent,
                passed = passed,
                timeSeconds = timeSeconds,
                questionResults = questionResults
            )
        }
    }
}

File: ExamForge-iOS/ExamForge/ViewModels/ResultViewModel.swift
import Foundation

@MainActor
final class ResultViewModel: ObservableObject {
    @Published var quizResult: QuizResult?

    private let repository: QuestionRepository

    init(repository: QuestionRepository) {
        self.repository = repository
    }

    func loadResult(sessionId: Int64) {
        Task {
            guard let session = await repository.getSessionById(id: sessionId) else { return }
            let answers = await repository.getAnswersForSession(sessionId: sessionId)

            var questionResults: [QuestionResult] = []
            for answer in answers {
                let question = await repository.getQuestionById(id: answer.questionId)
                questionResults.append(QuestionResult(
                    questionId:      answer.questionId,
                    questionText:    question?.questionText ?? "",
                    options:         question?.options ?? [],
                    correctIndices:  question?.correctAnswerIndices ?? [],
                    selectedIndices: answer.selectedIndices,
                    isCorrect:       answer.isCorrect,
                    explanation:     question?.explanation ?? ""
                ))
            }

            let total       = questionResults.count
            let correct     = questionResults.filter { $0.isCorrect }.count
            let scorePercent = total > 0 ? (correct * 100 / total) : 0
            let skipped     = questionResults.filter { $0.selectedIndices.isEmpty }.count

            quizResult = QuizResult(
                sessionId:       sessionId,
                totalQuestions:  total,
                correctAnswers:  correct,
                wrongAnswers:    total - correct - skipped,
                skippedAnswers:  skipped,
                scorePercent:    scorePercent,
                passed:          scorePercent >= 50,
                timeSeconds:     session.durationSeconds,
                questionResults: questionResults
            )
        }
    }
}
